File: 5.go
package main

import "fmt"

type pairOfNumbersReturningSomething func(int, int) int

type decisionFunctionPlaceholder func(pairOfNumbersReturningSomething) int

func cons(a, b int) decisionFunctionPlaceholder {
	thisReturnsThePair := func(somePair pairOfNumbersReturningSomething) int {
		return somePair(a, b)
	}

	return thisReturnsThePair
}

func car(fChoice decisionFunctionPlaceholder) int {
	thisReturnsTheFirstChoice := func(a, b int) int {
		return a
	}

	return fChoice(thisReturnsTheFirstChoice)
}

func cdr(lChoice decisionFunctionPlaceholder) int {
	thisReturnsTheSecondChoice := func(a, b int) int {
		return b
	}

	return lChoice(thisReturnsTheSecondChoice)
}

func main() {

	carVar := car(cons(1000, 9999))
	cdrVar := cdr(cons(1000, 9999))

	fmt.Println(carVar)
	fmt.Println(cdrVar)

}

File: 5.erl
-module(5).
-export([start/0]).

start() ->
    io:format("This is ~p~n", [car(1234, 5678)]),
    io:format("This is ~p~n", [cdr(1111, 9999)]).


pair(A, B) ->
    {A, B}.

cons(A, B) ->
    P = pair(A, B),
    P.

car(A, B) ->
    {FirstValue, _} = pair(A, B),
    FirstValue.

cdr(A, B) ->
    {_, SecondValue} = pair(A, B),
    SecondValue.